lib/portlayer/network/context.go

```diff
@@ -940,7 +940,13 @@ func (c *Context) RemoveIDFromScopes(op trace.Operation, id string) ([]*Endpoint
 		s := ne.scope
 
 		// save the endpoint info
-		e := con.Endpoint(s).copy()
+		e := con.Endpoint(s)
```

Comment:
I think there is something else going on here. This should not be `nil`. I would say we need to dig deeper; this just works around a bug in the code somewhere else.

Reply:
Well it is possible for `Endpoint(foo)` to return `nil`, which means we do have to handle that case, and we're not doing that here. I think you're right that there is likely more going on, but this would at least allow the PL to not panic & return a sane error along this code path. I'm still digging into it -- I think it's likely that this is not the whole fix, but I do think it's part of the fix at least.

Reply:
I mean I am fine with you putting this in to debug the problem, but I don't want to merge it like this. In this code path, it looks like this should not be nil. There is some race condition or some other assumption being made here that may or may not be correct. I just want us to figure that out first.

Reply:
Just looking at this code, I see that we are iterating over con.endpoints without acquiring a lock on con. con.Endpoint does acquire the lock. So we should at least fix that.